File: play-ws-standalone-json/src/main/scala/play/api/libs/ws/JsonBodyWritables.scala
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.libs.ws

import akka.util.ByteString
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import play.api.libs.json.JsValue
import play.api.libs.json.Json

trait JsonBodyWritables {

  /**
   * Creates an InMemoryBody with "application/json" content type, using the static ObjectMapper.
   */
  implicit val writeableOf_JsValue: BodyWritable[JsValue] = {
    BodyWritable(a => InMemoryBody(ByteString.fromArrayUnsafe(Json.toBytes(a))), "application/json")
  }

  def body(objectMapper: ObjectMapper): BodyWritable[JsonNode] =
    BodyWritable(
      json => InMemoryBody(ByteString.fromArrayUnsafe(objectMapper.writer.writeValueAsBytes(json))),
      "application/json"
    )
}

object JsonBodyWritables extends JsonBodyWritables

File: play-ws-standalone-json/src/main/java/play/libs/ws/JsonBodyWritables.java
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */

package play.libs.ws;

import org.apache.pekko.util.ByteString;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

public interface JsonBodyWritables {

    JsonBodyWritables instance = new JsonBodyWritables() {};

    /**
     * Creates a {@link InMemoryBodyWritable} for JSON, setting the content-type to "application/json", using the
     * default object mapper.
     *
     * @param node the node to pass in.
     * @return a {@link InMemoryBodyWritable} instance.
     */
    default BodyWritable<ByteString> body(JsonNode node) {
        return body(node, DefaultObjectMapper.instance);
    }

    /**
     * Creates a {@link InMemoryBodyWritable} for JSON, setting the content-type to "application/json".
     *
     * @param node the node to pass in.
     * @param objectMapper the object mapper to create a JSON document.
     * @return a {@link InMemoryBodyWritable} instance.
     */
    default BodyWritable<ByteString> body(JsonNode node, ObjectMapper objectMapper) {
        try {
            Object json = objectMapper.readValue(node.toString(), Object.class);
            byte[] bytes = objectMapper.writerWithDefaultPrettyPrinter().writeValueAsBytes(json);
            return new InMemoryBodyWritable(ByteString.fromArrayUnsafe(bytes), "application/json");
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }
}
